src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+for handling different events on the widget.  There's also a very nice
+`tutorial <https://facebook.github.io/react/tutorial/tutorial.html>`_ for
+building a tic-tac-toe game.  If you can grok that then you'll have the hang of
+React.
+
+If you've ever used a GUI toolkit like wx or Swing it shouldn't be too hard
+to pick up on what it's doing.
+
+Using React is quite a bit different from the old-fashioned way of making
+reactive web UIs with JavaScript.  What I'm calling the "old-fashioned" way is
+a couple things.  For one, the server might render serve up a bunch of HTML
+containing all the elements in your page, many of which might by "hidden" using
+CSS, and the JavaScript would hide and unhide elements on the page.  Or the
+JavaScript might generate some elements and insert them directly into the DOM
+and remove them as needed, either using the DOM API directly or, somewhat
+later, tools like `jQuery`_ (note: jQuery still has a role to play even in
+conjunction with React though).
+
+In other words, gone are the days of servers rendering and returning HTML to
+the browser.  All the rendering is pushed entirely to the client, with the
+client-server communication focused on as light-weight as possible message
+passing.  This potentially frees up enormous resources for the server, while
+pushing much more work to the client (which is why so many of your browser
+tabs are using over 100 MB of memory, among other reasons).
+
+The way React works, in short, is this: It maintains its own "virtual DOM"
+separate from the actual DOM of the browser document, with the same API as the
+real DOM.  Whenever you show, hide, or otherwise update the contents of a UI
+element in the application, it uses a copy of its virtual DOM to figure out
+exactly what needs to change in order for that to happen, and generates (and
+subsequently applies) a stream of operations to perform on the actual DOM in
+order to enact those changes.  The result is that there's nothing in the real
+DOM except for what's actually displayed on the page, which is convenient for
+debugging and inspection via your browser's development tools.  There's a
+simple `demonstration
+<https://facebook.github.io/react/docs/rendering-elements.html#react-only-updates-whats-necessary>`_
+of this aspect in the docs.
+
+Another nice aspect of React is its JSX domain-specific language which I'll
+discuss more next.
+
+As mentioned in the section on pug, essentially all of SMC's web frontend is
+built using React.  Almost no HTML is ever sent from the server.  Instead
+the frontend is built up by React.  When user interactions with the UI need
+to be persisted, those are sent as event messages (typically over
+WebSockets) to the server, which may in turn respond with events that result
+in updating the UI in some appropriate way (the event messages are usually a
+JSON object of some kind).  This is still an over-simplification (see for
+example the section on Redux later), but that's the basic idea.
+
```

Comment:
> My understanding though was that most of that has to do with the fact that you're using third-party tools, like codemirror.

No.  That is not correct at all.  The issue is simply that it is a lot of work.  React is basically "functional programming" and writing using that approach to program is more time consuming and harder.  Maintenance and bug fixing and reading later is easier.  

>  That seems pretty hard to get around. 

No -- it's orthogonal to the problem.   You can embed CodeMirror (or any other third party javascript widget) as a wrapped react component...   That's the very beginning of getting going on such a project.
 
> Are there any particular challenges that represents, other than some code not using the React interfaces?

Yes, many interesting challenges...  With react/redux/etc. you have to very carefully think through your data model, one way flow of data, etc.     This is related to your oversimplification about React just being "MVC with different names".  Not appreciating the subtlety of the problems doesn't make them go away, unfortunately.    As with all programming, from high up it all looks the same, but when you actually dive in to really do something, you start seeing the real problems, which are deep and numerous.